### ISSUE TITLE
Mergician Does Not Maintain All Properties

### ISSUE THREAD

**Issue by MichaelFoss** [jhildenbiddle/mergician#5]
When merging two objects, the original data is not always included in the resultant object.

Consider the following example:

```javascript
import mergician from 'mergician';

// See attached files for const values
import { input, translation } from './data.js';

const main = () => {
  // See attached file for what this looks like
  const result = mergician(input, translation);
  // This always exists
  console.log(input.events[0].decisions);
  // This SHOULD exist, but doesn't
  console.log(result.events[0].decisions);
};

main();
```

[data.js.input.json](https://github.com/jhildenbiddle/mergician/files/13732932/data.js.input.json)
[data.js.translation.json](https://github.com/jhildenbiddle/mergician/files/13732934/data.js.translation.json)
[data.js.result.json](https://github.com/jhildenbiddle/mergician/files/13732933/data.js.result.json)

Basically, `input` + `translation` = `result`, but `result` is missing `events[0].decisions` that is found in `input`. This is also true for `lands.shire.population`, `lands.shire.isFrodoHome`, `lands.rivendell.isSafeHaven`, `lands.rivendell.architecture.age`, `villians.saruman.powerLevel`, and `villians.sauron.powerLevel` – essentially every case where the original value has an object with a property that doesn't have the merged value with the corresponding object _doesn't_ have that property.

**Comment by jhildenbiddle** [comment]
Hi @MichaelFoss.

This is the correct behavior. Fortunately there is an easy way for you to solve your issue.

First, an explanation of the current behavior. 

Mergician does not merge objects within arrays because there is no way to know if objects within different arrays should be merged. Consider the following example:

```js
const obj1 = {
  data: [
    { a: 1 }
  ]
}

const obj2 = {
  data: [
    { b: 2 }
  ]
}
```

When Mergician merges `obj1` and `obj2`, the resulting object will look like this:

```js
 {
  data: [
    { a: 1 }
    { b: 2 }
  ]
}
```

You're asking Mergician to produce an an object like this:

```js
 {
  data: [
    { a: 1, b: 2 }
  ]
}
```

This seems easy enough to handle by adding a new `mergeObjectsInArrays` option to Mergician (similar to how `appendArrays` and `prependArrays` allow for merging array data). However, doing this successfully is dependent on two arrays having all related objects in the same order. If they aren't, a merge can complete successfully but produce corrupted data:

```js
const obj1 = {
  data: [
    { name: "John", age: 20 },
    { name: "Bill", age: 30 },
  ]
}

const obj2 = {
  data: [
    { name: "Bill", city: "Seattle" },
    { name: "John", city: "Austin" },
  ]
}
```

```js
 {
  data: [
    { name: "Bill", age: 20, city: "Seattle" }, // Name should be "John"
    { name: "John", age: 30, city: "Austin" }, // Name should be "Bill"
  ]
}
```

Fortunately, there are two paths forward:

1. I could add a `mergeObjectsInArrays` option and explicitly state in the documentation that this is a risky option that should only be used when arrays with objects to be merged are of the same length and in the same order.
2. You can use the [`beforeEach()`](https://jhildenbiddle.github.io/mergician/#/?id=beforeeach) to do essentially the same thing I would do with the `mergeObjectsInArrays` option checking to see if the values to merge are arrays and of the same length and contain at least one object, iterating over the arrays and merging each object individually, the returning that value as the final "merged" value.

Option 2 is the one I assumed people would use for the issue you're having. It's a more flexible approach because you can change behavior based on the key, value, nesting depth, etc. That said, I understand the desire to have something like a `mergeObjectsInArrays` for scenarios like yours where you are certain arrays with objects can and should be merged.

**Comment by jhildenbiddle** [comment]
Closing comment as resolved for now. Let me know if you would like to discuss further. Thanks!

**Comment by ptdev** [comment]
Hey, sorry for commenting on a closed issue, but I need something like this. Merge objects that contain arrays of objects to also be merged.

Can we get an example on how to implement this functionality using the beforeEach() method as you've hinted ?

Thanks.

**Comment by ptdev** [comment]
Hi, 

Quick followup here to answer my own question.

I've managed to achieve what I wanted by using the following:

```
const ret = mergician({    

        filter({ depth, key, srcObj, srcVal, targetObj, targetVal }) {

            if(key === 'prods') {
                if(targetObj['prods']) {
                    for(let i=0;i<srcVal.length;i++) {             
                        targetObj['prods'][i] = { ...targetObj['prods'][i], ...srcVal[i] }   
                    }
                } else {                    
                    targetObj['prods'] = srcVal
                }                
                return false
            }            

        }

})(common, transl)
```

Not sure if it's the best way to achieve this as I'm using the filter method instead of beforeEach as you've suggested, but I'm getting the correct results for my personal use case so that solves it for now.

Thanks for this lib, cheers.

**Comment by jhildenbiddle** [comment]
Hi @ptdev.

That works. Both the `filter` and `beforeEach` allow you to inspect and modify objects. The difference between the two options is their intended purpose (filter = include/exclude properties, beforeEach = inspect/modify values before merging).

There are a few potential issues in your solution worth considering:

1. You are shallowing merging `"prods"` data using [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax). This isn't necessarily a bad thing, but it is something to consider if  `"prods"` data contains nested properties today or could in the future.
2. You aren't doing checks on `"prods"` data before merging which could lead to errors or data corruption (see my previous [comment](https://github.com/jhildenbiddle/mergician/issues/5#issuecomment-1867061672) for a simple example).

Here is an alternative approach using `beforeEach` with a few benefits:

- Handles merging all objects in arrays (including nested arrays and arrays with mixed item types)
- Deep merges objects in arrays
- Includes basic data corruption checks (arrays must be of same length)

```js
const obj1 = {
  data: [
    'foo', 
    { a: 1, nested: [{ a: 1 }] }
  ],
};

const obj2 = {
  data: [
    'bar', 
    { b: 2, nested: [{ b: 2 }] }
  ],
};

const mergeObjectsInArrays = mergician({
  beforeEach({ srcVal, targetVal }) {
    const isArrayMerge =
      Array.isArray(srcVal) &&
      Array.isArray(targetVal) &&
      srcVal.length === targetVal.length;
    
    if (isArrayMerge) {
      const newArray = [];
      
      for (let i = 0; i < srcVal.length; i++) {
        const isObjectMerge = [srcVal[i], targetVal[i]].every((val) =>
          typeof val === 'object' &&
          val !== null &&
          !Array.isArray(val)
        );
          
        let val = srcVal[i];
          
        if (isObjectMerge) {
          val = mergeObjectsInArrays(targetVal[i], srcVal[i]);
        }
          
        newArray.push(val);
      }
        
      return newArray;
    }
  },
});
  
const results = mergeObjectsInArrays(obj1, obj2);
```

Result:

```js
{ 
  data: [
    'bar', 
    { a: 1, b: 2, nested: [ { a: 1, b: 2 } ] } 
  ] 
}
```

The "better" approach depends on what problem(s) you are trying to solve. The solution I've shared here is a generic solution that works on all arrays containing objects, which is similar to how I might implement a new `mergeObjectsInArrays: true|false` option for the library.

Hope this helps.

**Comment by alexwhitman** [comment]
Appreciate this is closed and labeled as `wontfix` but I'll throw in another voice in support of a `mergeObjectsInArrays: true|false` option.
